src/coprocessor/codec/mysql/decimal.rs

```diff
@@ -734,8 +732,8 @@ fn do_div_mod(
             res.frac_cnt = frac_word_to - int_word_to as u8;
             res = Res::Truncated(res.unwrap());
         }
-        for idx in l_idx..l_stop as usize {
-            res.word_buf[idx_to as usize] = buf[idx];
+        for v in buf.iter().take(l_stop as usize).skip(l_idx) {
```

Comment:
maybe the original code is more readable?

Reply:
I'm actually neutral to this clippy lint. But I agree the original code is more friendly to people with C/C++ background. So should we wait for other people to give some more advices on this or just revert it to the original code.

Reply:
any performance difference? 

seem we should be more rust-style :-)

Reply:
I really doubt if it could make significant difference, it is such a simple thing after all. 
The futures combinators is some what like this, If this makes noticeable difference, the futures library might suffer as well.

Reply:
Sorry that I was wrong. It did make significant difference. 

```rust
#![feature(test)]
extern crate test;

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;

    fn test1(array: &[i32]) -> i32 {
        let mut r = 0;
        for i in array.iter().take(99999).skip(90000) {
            r += i;
        }   
        r   
    }   

    fn test2(array: &[i32]) -> i32 {
        let mut r = 0;
        for idx in 90000..99999 {
            r += array[idx];
        }   
        r   
    }   

    #[bench]
    fn bench1(b: &mut Bencher) {
        let mut v = Vec::with_capacity(100000);
        v.resize(100000, 0); 
        b.iter(|| test1(&v));
    }   

    #[bench]
    fn bench2(b: &mut Bencher) {
        let mut v = Vec::with_capacity(100000);
        v.resize(100000, 0); 
        b.iter(|| test2(&v));
    }   
}
```

running 2 tests
test tests::bench1 ... bench:       9,750 ns/iter (+/- 330)
test tests::bench2 ... bench:       3,982 ns/iter (+/- 317)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out

Reply:
Hmmm. How about wrapping constants with a `test::black_box`? Maybe the compiler will optimize away some instructions in test2 because it's much more simpler.

Reply:
In case I misunderstood anything, I wrapped almost everything with test::black_box, the difference is less significant however still can not be ignored.
```
running 2 tests
test tests::bench1 ... bench:       8,599 ns/iter (+/- 1,357)
test tests::bench2 ... bench:       5,368 ns/iter (+/- 604)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out
```
Here is the new code
```rust
#![feature(test)]
extern crate test;

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;

    fn test1(array: &[i32]) -> i32 {
        let mut r = test::black_box(0);
        for i in array.iter().take(99999).skip(90000) {
            r += test::black_box(i);
        }
        test::black_box(r)
    }   

    fn test2(array: &[i32]) -> i32 {
        let mut r = test::black_box(0);
        for idx in 90000..99999 {
            r += test::black_box(array[idx]);
        }
        test::black_box(r)
    }   

    #[bench]
    fn bench1(b: &mut Bencher) {
        let mut v = Vec::with_capacity(100000);
        v.resize(100000, 0); 
        b.iter(|| test::black_box(test1(test::black_box(&v))));
    }   

    #[bench]
    fn bench2(b: &mut Bencher) {
        let mut v = Vec::with_capacity(100000);
        v.resize(100000, 0); 
        b.iter(|| test::black_box(test2(test::black_box(&v))));
    }   
}
```

Reply:
Oh. I also modified your code and made some tests by myself and it seems that there is notable difference in performance. I think we'd rather keep the old way. @siddontang 

Here is the result of adding black_box around the magic numbers:

```
test tests::bench1 ... bench:       8,968 ns/iter (+/- 1,313)
test tests::bench2 ... bench:       3,460 ns/iter (+/- 556)
```

Reply:
got it

Reply:
seem we need to send an issue to clippy :-)

Reply:
I figured out why the rust-style code is slower now. Looks like the compiler is smart enough to use the 64 bits register to run two 32 bits additions per iteration with the C/CPP style. The Rust style can perform only one computation at a time on the other hand. Here are generated assembly code snippets for two different implementation it is obvious that the CPP style version does two addl computation **(at _.Ltmp25_ and .Ltmp26_)** per loop. The Rust style implementation dose only one at ***.LBB0_9**

Assembly for **C/CPP** style implementation.
```asm
_ZN4test29test_function_cplusplus_style17hcf383ab4243ef9f2E:
.Lfunc_begin1:
	pushq	%rax
	movq	%rsi, %r8
	xorl	%eax, %eax
	movl	$90000, %edx
	movl	$90002, %ecx
.Ltmp22:
	.p2align	4, 0x90
.LBB1_1:
	leaq	-2(%rcx), %rsi
	cmpq	%r8, %rsi
	jae	.LBB1_5
.Ltmp23:
	leaq	-1(%rcx), %rsi
	cmpq	%r8, %rsi
	jae	.LBB1_3
.Ltmp24:
	cmpq	%r8, %rcx
	jae	.LBB1_7
.Ltmp25:
	addl	-8(%rdi,%rcx,4), %eax
.Ltmp26:
	addl	-4(%rdi,%rcx,4), %eax
.Ltmp27:
	addl	(%rdi,%rcx,4), %eax
.Ltmp28:
	addq	$3, %rdx
.Ltmp29:
	leaq	3(%rcx), %rsi
.Ltmp30:
	addq	$1, %rcx
	cmpq	$99999, %rcx
	movq	%rsi, %rcx
.Ltmp31:
	jb	.LBB1_1
.Ltmp32:
	popq	%rcx
	retq
.Ltmp33:
.LBB1_3:
	addq	$1, %rdx
	jmp	.LBB1_4
.Ltmp34:
.LBB1_7:
	addq	$2, %rdx
.Ltmp35:
.LBB1_4:
	movq	%rdx, %rsi
.Ltmp36:
.LBB1_5:
	leaq	.Lpanic_bounds_check_loc.4(%rip), %rdi
.Ltmp37:
	movq	%r8, %rdx
	callq	_ZN4core9panicking18panic_bounds_check17h9f8c971ae9e3c5dfE@PLT
.Ltmp38:
	ud2
.Ltmp39:
.Lfunc_end1:
```
Assembly for **Rust** style implementation.
```asm
_ZN4test24test_function_rust_style17he21fb9d884963274E:
.Lfunc_begin0:
	leaq	(%rdi,%rsi,4), %r8
	xorl	%eax, %eax
	movl	$90000, %esi
.Ltmp0:
	movl	$99999, %edx
.Ltmp1:
	testq	%rsi, %rsi
	jne	.LBB0_5
	jmp	.LBB0_2
.Ltmp2:
	.p2align	4, 0x90
.LBB0_9:
	addl	(%rdi), %eax
.Ltmp3:
	movq	%rsi, %rdi
.Ltmp4:
	xorl	%esi, %esi
.Ltmp5:
	testq	%rsi, %rsi
	je	.LBB0_2
.Ltmp6:
.LBB0_5:
	cmpq	%rsi, %rdx
	jbe	.LBB0_10
.Ltmp7:
	movq	%r8, %rcx
	subq	%rdi, %rcx
.Ltmp8:
	shrq	$2, %rcx
.Ltmp9:
	cmpq	%rsi, %rcx
	jbe	.LBB0_10
.Ltmp10:
	leaq	(%rdi,%rsi,4), %rcx
	orq	$1, %rsi
.Ltmp11:
	subq	%rsi, %rdx
.Ltmp12:
	leaq	(%rdi,%rsi,4), %rsi
.Ltmp13:
	movq	%rcx, %rdi
.Ltmp14:
	testq	%rdi, %rdi
	jne	.LBB0_9
	jmp	.LBB0_10
.Ltmp15:
	.p2align	4, 0x90
.LBB0_2:
	testq	%rdx, %rdx
	je	.LBB0_10
.Ltmp16:
	cmpq	%rdi, %r8
	je	.LBB0_10
.Ltmp17:
	addq	$-1, %rdx
.Ltmp18:
	movq	%rdi, %rsi
	addq	$4, %rsi
.Ltmp19:
	testq	%rdi, %rdi
	jne	.LBB0_9
.Ltmp20:
.LBB0_10:
	retq
.Ltmp21:
.Lfunc_end0:
```

Reply:
I did another test on i686 version which doesn't do this optimization, the result totally agrees that the reason was the two computation per iteration.
```
running 2 tests
test tests::bench1 ... bench:       8,035 ns/iter (+/- 876)
test tests::bench2 ... bench:      10,724 ns/iter (+/- 1,525)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out
```
And here are the generated assembly on i686.

There is only one addl at **.Ltmp29** for **C/CPP** style implementation.
```asm
_ZN4test29test_function_cplusplus_style17hcf383ab4243ef9f2E:
.Lfunc_begin1:
	pushl	%ebx
	pushl	%esi
	pushl	%eax
	movl	20(%esp), %edx
.Ltmp25:
	movl	16(%esp), %esi
.Ltmp26:
	calll	.L1$pb
.L1$pb:
	popl	%ebx
	movl	$90000, %ecx
	xorl	%eax, %eax
.Ltmp27:
	addl	$_GLOBAL_OFFSET_TABLE_+(.Ltmp27-.L1$pb), %ebx
.Ltmp28:
	.p2align	4, 0x90
.LBB1_1:
	cmpl	%edx, %ecx
	jae	.LBB1_4
.Ltmp29:
	addl	(%esi,%ecx,4), %eax
.Ltmp30:
	incl	%ecx
.Ltmp31:
	cmpl	$99999, %ecx
.Ltmp32:
	jb	.LBB1_1
.Ltmp33:
	addl	$4, %esp
.Ltmp34:
	popl	%esi
	popl	%ebx
	retl
.LBB1_4:
.Ltmp35:
	subl	$4, %esp
.Ltmp36:
	leal	.L__unnamed_1@GOTOFF(%ebx), %eax
.Ltmp37:
	pushl	%edx
	pushl	%ecx
	pushl	%eax
	calll	_ZN4core9panicking18panic_bounds_check17he66107f5e5f2fa8bE@PLT
.Ltmp38:
	addl	$16, %esp
	ud2
.Ltmp39:
.Lfunc_end1:
```

The assembly for **Rust** style implementation.
```asm
_ZN4test24test_function_rust_style17he21fb9d884963274E:
.Lfunc_begin0:
	pushl	%ebx
	pushl	%edi
	pushl	%esi
	movl	20(%esp), %eax
	movl	16(%esp), %esi
.Ltmp0:
	movl	$90000, %edi
	movl	$99999, %edx
.Ltmp1:
	leal	(%esi,%eax,4), %ecx
.Ltmp2:
	xorl	%eax, %eax
.Ltmp3:
	testl	%edi, %edi
	jne	.LBB0_5
	jmp	.LBB0_2
.Ltmp4:
	.p2align	4, 0x90
.LBB0_9:
	addl	(%esi), %eax
.Ltmp5:
	movl	%edi, %esi
.Ltmp6:
	xorl	%edi, %edi
.Ltmp7:
	testl	%edi, %edi
	je	.LBB0_2
.Ltmp8:
.LBB0_5:
	cmpl	%edi, %edx
	jbe	.LBB0_10
.Ltmp9:
	movl	%ecx, %ebx
	subl	%esi, %ebx
.Ltmp10:
	shrl	$2, %ebx
.Ltmp11:
	cmpl	%edi, %ebx
	jbe	.LBB0_10
.Ltmp12:
	movl	%edi, %ebx
	notl	%ebx
	addl	%ebx, %edx
.Ltmp13:
	leal	(%esi,%edi,4), %ebx
	shll	$2, %edi
.Ltmp14:
	orl	$4, %edi
	addl	%esi, %edi
.Ltmp15:
	movl	%ebx, %esi
.Ltmp16:
	testl	%esi, %esi
	jne	.LBB0_9
	jmp	.LBB0_10
.Ltmp17:
	.p2align	4, 0x90
.LBB0_2:
	testl	%edx, %edx
	je	.LBB0_10
.Ltmp18:
	cmpl	%esi, %ecx
	je	.LBB0_10
.Ltmp19:
	movl	%esi, %edi
.Ltmp20:
	decl	%edx
.Ltmp21:
	addl	$4, %edi
.Ltmp22:
	testl	%esi, %esi
	jne	.LBB0_9
.Ltmp23:
.LBB0_10:
	popl	%esi
	.cfi_def_cfa_offset 12
	popl	%edi
	.cfi_def_cfa_offset 8
	popl	%ebx
	.cfi_def_cfa_offset 4
	retl
.Ltmp24:
```

Reply:
> seem we need to send an issue to clippy :-)

@siddontang  imo this codegen issue should be sent to rustc :)

BTW why not use `slice::copy_from_slice`?

Reply:
Hm, after a second thought, it actually doesn't use 64 bits register. It just simply read 32 bits data twice. It should be able to do similar thing in i686 version.